TestDockerStart - `panic: Log in goroutine after TestDockerStart has completed`

## Flaky Test

* **Test Name:** TestDockerStart
* ~~**Link:**~~
* **Branch:** 9.3
* **Artifact Link:** https://buildkite.com/elastic/beats-libbeat/builds/29269/canvas?jid=019e02c9-853c-457f-ada3-ed4ce712d8b7&tab=output#L432
* ~~**Notes:**~~

### Stack Trace

```
=== Failed
=== FAIL: libbeat/autodiscover/providers/docker  (0.00s)
PASS
panic: Log in goroutine after TestDockerStart has completed: 2026-05-07T14:34:18.278Z	DEBUG	docker.docker	Got a new docker event: {destroy 03d3ca6bf4730e32ff200d3c8a2a145d2599b7160d4fc8e1f0abc50115c1ac71 busybox:latest container destroy {03d3ca6bf4730e32ff200d3c8a2a145d2599b7160d4fc8e1f0abc50115c1ac71 map[image:busybox:latest label:foo label.child:bar name:festive_bhabha]} local 1778164458 1778164458275471000}

```

--- a/libbeat/autodiscover/providers/docker/docker.go
+++ b/libbeat/autodiscover/providers/docker/docker.go
@@ -23,6 +23,7 @@ import (
 	"errors"
 	"fmt"
 	"strconv"
+	"sync"
 	"time"
 
 	"github.com/gofrs/uuid/v5"
@@ -61,6 +62,7 @@ type Provider struct {
 	stoppers      map[string]*time.Timer
 	stopTrigger   chan *dockerContainerMetadata
 	logger        *logp.Logger
+	stopWg        sync.WaitGroup
 }
 
 // AutodiscoverBuilder builds and returns an autodiscover provider
@@ -129,17 +131,19 @@ func AutodiscoverBuilder(
 		stoppers:      make(map[string]*time.Timer),
 		stopTrigger:   make(chan *dockerContainerMetadata),
 		logger:        logger,
+		stopWg:        sync.WaitGroup{},
 	}, nil
 }
 
 // Start the autodiscover process
 func (d *Provider) Start() {
-	go func() {
+	d.stopWg.Go(func() {
 		for {
 			select {
 			case <-d.stop:
 				d.startListener.Stop()
 				d.stopListener.Stop()
+				d.watcher.Stop()
 
 				// Stop all timers before closing the channel
 				for _, stopper := range d.stoppers {
@@ -158,7 +162,7 @@ func (d *Provider) Start() {
 				d.stopContainer(target.container, target.metadata)
 			}
 		}
-	}()
+	})
 }
 
 type dockerContainerMetadata struct {
@@ -399,6 +403,7 @@ func (d *Provider) generateHints(event bus.Event) bus.Event {
 // Stop the autodiscover process
 func (d *Provider) Stop() {
 	close(d.stop)
+	d.stopWg.Wait()
 }
 
 func (d *Provider) String() string {

